include/net/http_app.h

```diff
@@ -0,0 +1,1079 @@
+ */
+int http_client_send_req(struct http_ctx *http_ctx,
+			 struct http_request *req,
+			 http_response_cb_t cb,
+			 u8_t *response_buf,
+			 size_t response_buf_len,
+			 void *user_data,
+			 s32_t timeout);
+
+/**
+ * @brief Send a HTTP GET request to peer.
+ *
+ * @param http_ctx HTTP context.
+ * @param url URL to use.
+ * @param host Host field in HTTP header. If set to NULL, then server
+ * name is used.
+ * @param extra_header_fields Any extra header fields that caller wants
+ * to add. This can be set to NULL. The format is "name: value\r\n"
+ * Example: "Accept: text/plain\r\nConnection: Close\r\n"
+ * @param cb Callback to call when the response has been received from peer.
+ * @param response_buf Caller-supplied buffer where the HTTP request will be
+ * stored
+ * @param response_buf_len Length of the caller-supplied buffer.
+ * @param user_data A valid pointer on some user data or NULL
+ * @param timeout Amount of time to wait for a reply. If the timeout is 0,
+ * then we return immediately and the callback (if set) will be called later.
+ *
+ * @return Return 0 if ok, and <0 if error.
+ */
+static inline int http_client_send_get_req(struct http_ctx *http_ctx,
+					   const char *url,
+					   const char *host,
+					   const char *extra_header_fields,
+					   http_response_cb_t cb,
+					   u8_t *response_buf,
+					   size_t response_buf_len,
+					   void *user_data,
+					   s32_t timeout)
+{
+	struct http_request req = {
+		.method = HTTP_GET,
+		.url = url,
+		.host = host,
+		.protocol = " " HTTP_PROTOCOL HTTP_CRLF,
+		.header_fields = extra_header_fields,
+	};
+
+	return http_client_send_req(http_ctx, &req, cb, response_buf,
+				    response_buf_len, user_data, timeout);
+}
+
+/**
+ * @brief Send a HTTP POST request to peer.
+ *
+ * @param http_ctx HTTP context.
+ * @param url URL to use.
+ * @param host Host field in HTTP header. If set to NULL, then server
+ * name is used.
+ * @param extra_header_fields Any extra header fields that caller wants
+ * to add. This can be set to NULL. The format is "name: value\r\n"
+ * Example: "Accept: text/plain\r\nConnection: Close\r\n"
+ * @param content_type Content type of the data.
+ * @param payload Payload data.
+ * @param cb Callback to call when the response has been received from peer.
+ * @param response_buf Caller-supplied buffer where the HTTP response will be
+ * stored
+ * @param response_buf_len Length of the caller-supplied buffer.
+ * @param user_data A valid pointer on some user data or NULL
+ * @param timeout Amount of time to wait for a reply. If the timeout is 0,
+ * then we return immediately and the callback (if set) will be called later.
+ *
+ * @return Return 0 if ok, and <0 if error.
+ */
+static inline int http_client_send_post_req(struct http_ctx *http_ctx,
+					    const char *url,
+					    const char *host,
+					    const char *extra_header_fields,
+					    const char *content_type,
+					    const char *payload,
```

Comment:
(if you want to continue POST discussion). Here you effectively assume that POST payload is textual data. Why so? What's so wrong with posting binary file for example? (Yeah, adding yet another param will make signature even more cumbersome, but "stray" params there are `http_response_cb_t cb`, `user_data` (perhaps also response_buf - just malloc() it away, lol). Perhaps, we don't need these wrappers at all, and users should use `struct http_request` directly.